src/webgpu/api/operation/copyBetweenLinearDataAndTexture.spec.ts

```diff
@@ -0,0 +1,742 @@
+      dimension,
+      mipLevelCount: mipLevel + 1,
+      usage: GPUTextureUsage.COPY_SRC | GPUTextureUsage.COPY_DST,
+    });
+
+    const data = this.generateData(dataSize);
+
+    switch (checkMethod) {
+      case 'PartialCopyT2B': {
+        this.initTexture(
+          { texture, mipLevel, origin },
+          textureDataLayout,
+          copySize,
+          data,
+          initMethod
+        );
+
+        this.checkData({ texture, mipLevel, origin }, textureDataLayout, format, copySize, data);
+
+        break;
+      }
+      case 'FullCopyT2B': {
+        const fullTextureCopyLayout = getTextureCopyLayout(format, dimension, textureSize, {
+          mipLevel,
+        });
+
+        const fullData = this.getFullData({ texture, mipLevel }, fullTextureCopyLayout);
+
+        this.initTexture(
+          { texture, mipLevel, origin },
+          textureDataLayout,
+          copySize,
+          data.slice(0, dataSize),
+          initMethod
+        );
+
+        this.fullCheck(
+          { texture, mipLevel, origin },
+          fullTextureCopyLayout,
+          textureDataLayout,
+          copySize,
+          format,
+          fullData,
+          data
+        );
+
+        break;
+      }
+      default:
+        unreachable();
+    }
+  }
+}
+
+// This is a helper function used for filtering test parameters
+// TODO: Modify this after introducing tests with rendering.
+function formatCanBeTested({ format }: { format: SizedTextureFormat }): boolean {
+  return kSizedTextureFormatInfo[format].copyDst && kSizedTextureFormatInfo[format].copySrc;
+}
+
+const kMethodsToTest = [
+  { initMethod: 'WriteTexture', checkMethod: 'PartialCopyT2B' },
+  { initMethod: 'WriteTexture', checkMethod: 'FullCopyT2B' },
+  { initMethod: 'CopyB2T', checkMethod: 'FullCopyT2B' },
+] as const;
+
+export const g = makeTestGroup(CopyBetweenLinearDataAndTextureTest);
+
+// Test that copying data with various bytesPerRow and rowsPerImage values and minimum required
+// bytes in copy works for every format.
+// Covers a special code path for Metal:
+//    bufferSize - offset < bytesPerImage * copyExtent.depth
+// Covers a special code path for D3D12:
+//    when bytesPerRow is not a multiple of 512 and copyExtent.depth > 1: copyExtent.depth % 2 {==, >} 0
+//    bytesPerRow == bytesInACompleteCopyImage
+g.test('copy_with_various_rows_per_image_and_bytes_per_row')
+  .params(
+    params()
+      .combine(kMethodsToTest)
+      .combine([
+        { bytesPerRowPadding: 0, rowsPerImagePaddingInBlocks: 0 }, // no padding
+        { bytesPerRowPadding: 0, rowsPerImagePaddingInBlocks: 6 }, // rowsPerImage padding
+        { bytesPerRowPadding: 6, rowsPerImagePaddingInBlocks: 0 }, // bytesPerRow padding
+        { bytesPerRowPadding: 15, rowsPerImagePaddingInBlocks: 17 }, // both paddings
+      ])
+      .combine([
+        { copyWidthInBlocks: 3, copyHeightInBlocks: 4, copyDepth: 5 }, // standard copy, covers bytesPerRow = 256 and copyDepth % 2 == 1.
```

Comment:
nit: explains that this gives coverage of special cases in the D3D12 backend.

Reply:
Done